File: desktop/utils/retry.py
"""Retry mit Exponential-Backoff + Circuit-Breaker (nur Stdlib).

// REAL-IMPLEMENTATION 2026-09-11 (Phase 3): Gegenstück zu `src/lib/retry.ts`
für die Desktop-Konsole. Wiederholt nur transiente Fehler (Verbindung,
Timeout), niemals Anwendungsfehler. Spiegeldatei: `mobile-server/retry_util.py`.
"""
from __future__ import annotations

import random
import threading
import time
import urllib.error
from typing import Any, Callable, TypeVar

T = TypeVar("T")

TRANSIENT = (urllib.error.URLError, OSError, TimeoutError)


def with_retry(
    fn: Callable[[], T],
    *,
    retries: int = 2,
    base_delay: float = 0.3,
    max_delay: float = 4.0,
    exceptions: tuple[type[BaseException], ...] = TRANSIENT,
) -> T:
    """Führt ``fn`` aus, wiederholt transiente Fehler mit Backoff + Jitter."""
    last: BaseException = RuntimeError("ohne_versuch")
    for attempt in range(retries + 1):
        try:
            return fn()
        except exceptions as exc:
            last = exc
            if attempt >= retries:
                break
            delay = min(max_delay, base_delay * (2 ** attempt)) + random.uniform(0, base_delay * 0.5)
            time.sleep(delay)
    raise last


class CircuitBreaker:
    """Fehlerzähler je Gegenstelle: nach Dauerfehlern sofort abbrechen."""

    def __init__(self, fail_threshold: int = 5, reset_timeout: float = 30.0) -> None:
        self._fail_threshold = fail_threshold
        self._reset_timeout = reset_timeout
        self._failures = 0
        self._opened_at = 0.0
        self._lock = threading.Lock()

    @property
    def is_open(self) -> bool:
        with self._lock:
            if self._opened_at <= 0:
                return False
            if time.monotonic() - self._opened_at >= self._reset_timeout:
                self._opened_at = 0.0
                self._failures = 0
                return False
            return True

    def allow(self) -> bool:
        return not self.is_open

    def record_success(self) -> None:
        with self._lock:
            self._failures = 0
            self._opened_at = 0.0

    def record_failure(self) -> None:
        with self._lock:
            self._failures += 1
            if self._failures >= self._fail_threshold:
                self._opened_at = time.monotonic()

    def retry_in_s(self) -> float:
        with self._lock:
            if self._opened_at <= 0:
                return 0.0
            return max(0.0, self._reset_timeout - (time.monotonic() - self._opened_at))


_breakers: dict[str, CircuitBreaker] = {}
_breakers_lock = threading.Lock()
# Phase 5: Registry bleibt begrenzt (FIFO-Verdrängung, kein Leak).
MAX_BREAKERS = 128


def get_breaker(key: str, **kwargs: Any) -> CircuitBreaker:
    with _breakers_lock:
        breaker = _breakers.get(key)
        if breaker is None:
            breaker = CircuitBreaker(**kwargs)
            if len(_breakers) >= MAX_BREAKERS:
                _breakers.pop(next(iter(_breakers)))
            _breakers[key] = breaker
        return breaker


def reset_breakers() -> None:
    """Nur für Tests: alle Breaker zurücksetzen."""
    with _breakers_lock:
        _breakers.clear()

File: src/lib/retry.ts
/**
 * Retry mit Exponential-Backoff + Circuit-Breaker für externe Calls.
 *
 * // REAL-IMPLEMENTATION 2026-09-11 (Phase 3):
 * Zentrale Bausteine für alle Bridge-/Gateway-Aufrufe (`mcpClient`,
 * `portview`). Wiederholt nur transiente Netzwerkfehler (Timeout, Verbindungs-
 * abbruch), niemals Anwendungsfehler (HTTP 4xx/5xx-Payloads, JSON-Fehler).
 */

export interface RetryOptions {
  /** Versuche zusätzlich zum Erstversuch (Default 2). */
  retries?: number;
  /** Basiswartezeit in ms (Default 300), exponentiell wachsend + Jitter. */
  baseDelayMs?: number;
  /** Maximale Wartezeit zwischen Versuchen in ms (Default 4000). */
  maxDelayMs?: number;
  /** Timeout je Versuch in ms (Default 8000). */
  timeoutMs?: number;
}

export function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => {
    setTimeout(resolve, ms);
  });
}

function isTransient(error: unknown): boolean {
  // TypeError = Netzwerkfehler (offline, refused, CORS); TimeoutError/Abort =
  // unser AbortSignal.timeout bzw. abgebrochene Verbindung.
  if (error instanceof TypeError) return true;
  const name = (error as Error)?.name ?? '';
  return name === 'TimeoutError' || name === 'AbortError';
}

/**
 * `fetch` mit Timeout je Versuch und Exponential-Backoff bei transienten
 * Fehlern. Wirft den letzten Fehler, wenn alle Versuche scheitern.
 */
export async function fetchWithRetry(
  input: string,
  init: RequestInit = {},
  opts: RetryOptions = {},
): Promise<Response> {
  const retries = opts.retries ?? 2;
  const baseDelay = opts.baseDelayMs ?? 300;
  const maxDelay = opts.maxDelayMs ?? 4000;
  const timeoutMs = opts.timeoutMs ?? 8000;
  let lastError: unknown = new Error('fetch_ohne_versuch');
  for (let attempt = 0; attempt <= retries; attempt += 1) {
    try {
      const res = await fetch(input, {
        ...init,
        signal: AbortSignal.timeout(timeoutMs),
      });
      return res;
    } catch (e) {
      lastError = e;
      if (!isTransient(e) || attempt >= retries) break;
      const backoff = Math.min(maxDelay, baseDelay * 2 ** attempt);
      const jitter = Math.random() * baseDelay * 0.5;
      await sleep(backoff + jitter);
    }
  }
  throw lastError;
}

// ---------------------------------------------------------------------------
// Circuit-Breaker: schützt vor Dauerbeschuss toter Dienste
// ---------------------------------------------------------------------------

type BreakerState = 'closed' | 'open' | 'half-open';

export interface BreakerOptions {
  /** Fehler in Folge bis zum Öffnen (Default 5). */
  failThreshold?: number;
  /** Öffnungsdauer in ms bis zum Halb-Öffnen (Default 30 000). */
  resetTimeoutMs?: number;
}

export class CircuitBreaker {
  private state: BreakerState = 'closed';
  private failures = 0;
  private openedAt = 0;
  private readonly failThreshold: number;
  private readonly resetTimeoutMs: number;

  constructor(opts: BreakerOptions = {}) {
    this.failThreshold = opts.failThreshold ?? 5;
    this.resetTimeoutMs = opts.resetTimeoutMs ?? 30_000;
  }

  get current(): BreakerState {
    if (this.state === 'open' && Date.now() - this.openedAt >= this.resetTimeoutMs) {
      this.state = 'half-open';
    }
    return this.state;
  }

  /** Darf der Aufruf raus? (Nein → sofort strukturiert abbrechen.) */
  allow(): boolean {
    return this.current !== 'open';
  }

  recordSuccess(): void {
    this.failures = 0;
    this.state = 'closed';
  }

  recordFailure(): void {
    this.failures += 1;
    if (this.state === 'half-open' || this.failures >= this.failThreshold) {
      this.state = 'open';
      this.openedAt = Date.now();
    }
  }

  /** Millisekunden bis zum nächsten Halb-Öffnen (für Fehlermeldungen). */
  retryInMs(): number {
    if (this.current !== 'open') return 0;
    return Math.max(0, this.resetTimeoutMs - (Date.now() - this.openedAt));
  }
}

const breakers = new Map<string, CircuitBreaker>();
/** Phase 5: Registry bleibt begrenzt (FIFO-Verdrängung, kein Leak). */
const MAX_BREAKERS = 128;

/** Breaker je Gegenstelle (z. B. Origin der Bridge/des Gateways). */
export function getCircuitBreaker(key: string, opts: BreakerOptions = {}): CircuitBreaker {
  let breaker = breakers.get(key);
  if (!breaker) {
    breaker = new CircuitBreaker(opts);
    if (breakers.size >= MAX_BREAKERS) {
      const oldest = breakers.keys().next().value as string | undefined;
      if (oldest !== undefined) breakers.delete(oldest);
    }
    breakers.set(key, breaker);
  }
  return breaker;
}

/** Nur für Tests: alle Breaker zurücksetzen. */
export function resetCircuitBreakers(): void {
  breakers.clear();
}
